techdebt: [CDS-83153]: Revamp PR Pipelines - technical changes (#4352)
harness/developer-hub: docs/continuous-delivery/gitops/pr-pipelines/pr-pipelines-basics.md

---
title: Harness GitOps PR pipelines basics
description: Change an application in just one of its target environments.
sidebar_position: 1
---

:::note

Currently, this feature is behind the feature flag `NG_SVC_ENV_REDESIGN`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.

:::

This topic introduces you to Harness PR pipelines, which provides a first-class support for modifying existing GitOps Applications, especially for those that are generated by ApplicationSets using the [Git Generator](https://argocd-applicationset.readthedocs.io/en/stable/Generators-Git/).

Often, even though your ApplicationSet syncs one microservice/application to multiple target environments, you might want to change a microservice in just one of the target environments, such as a dev environment. A Harness PR Pipeline enables you to do this.

:::tip Important

> This topic builds on the ApplicationSet concept discussed in [Harness GitOps ApplicationSets](/docs/continuous-delivery/gitops/applicationsets/harness-git-ops-application-set-tutorial.md). It is recommended that you go through that topic before venturing into this one.

Reviewer: ```suggestion
This topic builds on the ApplicationSet concept discussed in [Harness GitOps ApplicationSets](/docs/continuous-delivery/gitops/applicationsets/harness-git-ops-application-set-tutorial.md). Please review that topic before learning about PR pipeline basics.
```

:::

A typical GitOps Application syncs a source manifest to a destination cluster. If you have multiple target clusters, you could create separate GitOps Applications for each one, but that makes management more challenging. Also, what if you want to sync an application with 100s of target clusters? Managing 100s of GitOps Applications is not acceptable.

To solve this use case, Harness supports ApplicationSets.

## ApplicationSets: A Quick Recap

An ApplicationSet can be thought of as a kind of Application factory. It defines one application template and syncs it to multiple target environments. It is similar to an Application but uses a template to achieve application automation with multiple target environments.

> You can add an ApplicationSet manifest to a Harness GitOps Application just like you would add a typical Deployment manifest. At runtime, Harness uses the ApplicationSet to deploy the applications generated by the common template to all the target environments' clusters.

Reviewer: ```suggestion
You can add an ApplicationSet manifest to a Harness GitOps Application just like you would add a typical deployment manifest. At runtime, Harness uses the ApplicationSet to deploy the applications generated by the common template to all the target environments' clusters.
```

![](static/harness-git-ops-application-set-tutorial-61.png)

ApplicationSets use generators to generate parameters that are substituted into the `template:` section of the ApplicationSet resource during template rendering.

> There are many types of generators, however, Harness provides first-class support for the **Git Generator** with the **Update Release Repo**, **Merge PR** and **Revert PR** [steps](/docs/continuous-delivery/gitops/use-gitops/gitops-pipeline-steps.md) in the PR pipeline.

Reviewer: ```suggestion
There are many types of generators, and Harness provides first-class support for the **Git Generator** with the **Update Release Repo**, **Merge PR**, and **Revert PR** [steps](/docs/continuous-delivery/gitops/use-gitops/gitops-pipeline-steps.md) in the PR pipeline.
```

Reviewer: Note that we use the oxford comma.

For more information on other Generators, please refer to [Generators](https://argocd-applicationset.readthedocs.io/en/stable/Generators/) from Argo CD docs.

## Git Generator

> At its core, ArgoCD's Git generator allows for dynamic creation of ArgoCD Application resources from templates using parameters sourced from a Git repository. This enables applications to be defined once in a template and instantiated multiple times with different configurations. The key aspect lies in its ability to pull environment-specific or application-specific configurations directly from a Git repository, ensuring version control and traceability.

Reviewer: ```suggestion
At its core, ArgoCD's Git generator allows for dynamic creation of ArgoCD Application resources from templates using parameters sourced from a Git repository. This functionality enables applications to be defined once in a template and instantiated multiple times with different configurations. The Git generator ensures version control and traceability by pulling environment or application-specific configurations directly from a Git repository.
```

## Harness PR Pipelines

Harness PR Pipelines allow you to manage the configuration of individual applications created by the ApplicationSet. The key differentiator between various config files for the ApplicationSet is the folder structure, which would usually differentiate based on the environment in Harness.

When you deploy a Harness PR Pipeline, you simply indicate what target environment application you want to update and the `config.json` keys/values you want changed, such as release tags. Harness creates the pull request in your Git repo and merges it for you. Now, the target environment application has the new keys/values, which can then be reviewed and deployed by syncing the ApplicationSet.

![](static/harness-git-ops-application-set-tutorial-62.png)

:::note

To create PR pipelines in Harness, please refer to [Create Harness GitOps PR pipelines](/docs/continuous-delivery/gitops/pr-pipelines/pr-pipelines.md).

:::

<details>

> <summary>An aside: Wave deployments</summary>

Reviewer: ```suggestion
<summary>Wave deployments</summary>
```
